### PR TITLE
Add Corrade 2020.06

### PR DESCRIPTION
**Pull request by helmesjo**
Specify library name and version:  **corrade/2020.06**

- Add v2020.06
- Fix `To use Corrade with MSVC 20XX, build it with MSVC20XX_COMPATIBILITY enabled` issue for consumers.
- Fixed so that consumers get the cmake scripts provided by corrade (used to read configs and setup correct preprocessor defines).

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.

### CODE REVIEW

**Comment by danimtb** [comment]
@helmesjo has you requested access in issue #4? Otherwise, the CI won't be able to run

**Comment by helmesjo** [comment]
@danimtb Thanks, just did that!

**Comment by helmesjo** [comment]
Still waiting for early access.

File: recipes/corrade/all/conanfile.py [hunk 1]
```diff
@@ -25,16 +24,18 @@ class CorradeConan(ConanFile):
         "with_pluginmanager": [True, False],
         "with_testsuite": [True, False],
         "with_utility": [True, False],
+        "with_rc": [True, False],
     }
     default_options = {
         "shared": False,
         "fPIC": True,
-        "build_deprecated": True,
+        "build_deprecated": False,
```
Comment (uilianries) [inline recipes/corrade/all/conanfile.py:32]:
```suggestion
```

Build deprecated is invalid, you can remove this option.

File: recipes/corrade/all/conanfile.py [hunk 2]
```diff
@@ -58,24 +59,28 @@ def source(self):
...
-            self._cmake.definitions["WITH_INTERCONNECT"] = self.options["with_interconnect"]
-            self._cmake.definitions["WITH_MAIN"] = self.options["with_main"]
-            self._cmake.definitions["WITH_PLUGINMANAGER"] = self.options["with_pluginmanager"]
-            self._cmake.definitions["WITH_TESTSUITE"] = self.options["with_testsuite"]
-            self._cmake.definitions["WITH_UTILITY"] = self.options["with_utility"]
-            self._cmake.definitions["WITH_RC"] = "ON"
+            def add_cmake_option(option, value):
+                var_name = "{}".format(option).upper()
+                value_str = "{}".format(value)
+                var_value = "ON" if value_str == 'True' else "OFF" if value_str == 'False' else value_str
```
Comment (uilianries) [inline recipes/corrade/all/conanfile.py:65]:
```suggestion
                var_value = bool(value)
```

File: recipes/corrade/all/conanfile.py [hunk 3]
```diff
@@ -58,24 +59,28 @@ def source(self):
...
-            self._cmake.definitions["WITH_PLUGINMANAGER"] = self.options["with_pluginmanager"]
-            self._cmake.definitions["WITH_TESTSUITE"] = self.options["with_testsuite"]
-            self._cmake.definitions["WITH_UTILITY"] = self.options["with_utility"]
-            self._cmake.definitions["WITH_RC"] = "ON"
+            def add_cmake_option(option, value):
+                var_name = "{}".format(option).upper()
+                value_str = "{}".format(value)
+                var_value = "ON" if value_str == 'True' else "OFF" if value_str == 'False' else value_str 
+                self._cmake.definitions[var_name] = var_value
+                print("{0}={1}".format(var_name, var_value))
```
Comment (uilianries) [inline recipes/corrade/all/conanfile.py:67]:
```suggestion
```

File: recipes/corrade/all/conanfile.py [hunk 4]
```diff
@@ -58,24 +59,28 @@ def source(self):
...
+
+            for option, value in self.options.items():
+                add_cmake_option(option, value)
+
+            add_cmake_option("BUILD_STATIC", not self.options.shared)
+            add_cmake_option("BUILD_STATIC_PIC", not self.options.shared and self.options.get_safe("fPIC") == True)
```
Comment (uilianries) [inline recipes/corrade/all/conanfile.py:73]:
```suggestion
            self._cmake.definitions["BUILD_STATIC"] = not self.options.shared
            self._cmake.definitions["BUILD_STATIC_PIC"] not self.options.shared and self.options.get_safe("fPIC", False)
```

File: recipes/corrade/all/conanfile.py [hunk 5]
```diff
@@ -58,24 +59,28 @@ def source(self):
...
+
+            if self.settings.compiler == "Visual Studio":
+                self._cmake.definitions["MSVC2015_COMPATIBILITY"] = "ON" if self.settings.compiler.version == "14" else "OFF"
+                self._cmake.definitions["MSVC2017_COMPATIBILITY"] = "ON" if self.settings.compiler.version == "15" else "OFF"
+                self._cmake.definitions["MSVC2019_COMPATIBILITY"] = "ON" if self.settings.compiler.version == "16" else "OFF"
```
Comment (uilianries) [inline recipes/corrade/all/conanfile.py:78]:
```suggestion
                self._cmake.definitions["MSVC2015_COMPATIBILITY"] = self.settings.compiler.version
                self._cmake.definitions["MSVC2017_COMPATIBILITY"] = self.settings.compiler.version
                self._cmake.definitions["MSVC2019_COMPATIBILITY"] = self.settings.compiler.version == "16"
```

File: recipes/corrade/all/conanfile.py [hunk 6]
```diff
@@ -90,8 +95,11 @@ def package(self):
         cmake.install()
 
         share_cmake = os.path.join(self.package_folder, "share", "cmake", "Corrade")
-        self.copy("UseCorrade.cmake", src=share_cmake, dst=os.path.join(self.package_folder, "lib", "cmake", "Corrade"))
+        self.copy("CMakeLists.txt", src=share_cmake, dst=os.path.join(self.package_folder, "lib", "cmake", "Corrade"))
+        self.copy("CorradeConfig.cmake", src=share_cmake, dst=os.path.join(self.package_folder, "lib", "cmake", "Corrade"))
         self.copy("CorradeLibSuffix.cmake", src=share_cmake, dst=os.path.join(self.package_folder, "lib", "cmake", "Corrade"))
+        self.copy("FindCorrade.cmake", src=share_cmake, dst=os.path.join(self.package_folder, "lib", "cmake", "Corrade"))
```
Comment (uilianries) [inline recipes/corrade/all/conanfile.py:101]: What specific macro do you need from Corrade? Conan CmakeFindPackage generator should be enough. If there is no specific macro there, please remove FindCorrade/CorradeConfig

File: recipes/corrade/all/conanfile.py [hunk 7]
```diff
@@ -136,8 +148,10 @@ def package_info(self):
...
-        bindir = os.path.join(self.package_folder, "bin")
-        self.output.info("Appending PATH environment variable: {}".format(bindir))
-        self.env_info.PATH.append(bindir)
+        self.env_info.PATH.append(os.path.join(self.package_folder, "bin"))
+        if self.options.shared:
+            # Help linker find rpaths
+            if self.settings.os == "Linux":
+                self.env_info.LD_LIBRARY_PATH.append(os.path.join(self.package_folder, "lib"))
+            if self.settings.os == "Macos":
+                self.env_info.DYLD_LIBRARY_PATH.append(os.path.join(self.package_folder, "lib"))
```
Comment (uilianries) [inline recipes/corrade/all/conanfile.py:157]:
```suggestion
```

Conan already manipulates it: https://docs.conan.io/en/latest/devtools/create_installer_packages.html#using-the-tool-packages-in-other-recipes

File: recipes/corrade/all/conanfile.py (hunk 4 above)
Comment (helmesjo) [reply recipes/corrade/all/conanfile.py:73]: Why not reuse the `add_cmake_option(...)` method?

File: recipes/corrade/all/conanfile.py (hunk 5 above)
Comment (helmesjo) [reply recipes/corrade/all/conanfile.py:78]: What happened to `== "14"` and `== "15"`?

File: recipes/corrade/all/conanfile.py (hunk 6 above)
Comment (helmesjo) [reply recipes/corrade/all/conanfile.py:101]: All the different options passed to CMake when building corrade (that are relevant for consumers) gets written to a `configure.h` which is packaged along all other includes. This file [is then parsed by `FindCorrade.cmake`](https://github.com/mosra/corrade/blob/162d6a7d77e4940f7c05613013e1d3aa522bc8c6/modules/FindCorrade.cmake#L276-L318). The alternative is to extract the necessary pieces and add to a CMake script that's part of the recipe (if I'm correct it's only the parsing of `configure.h` that's relevant for consumers).

Or alternatively mimic it by directly in `package_info()` setup all these different preprocessor directives. Though this is even more error prone than above, IMO.

File: recipes/corrade/all/conanfile.py (hunk 7 above)
Comment (helmesjo) [reply recipes/corrade/all/conanfile.py:157]: Pretty sure I had to add this at some point but can't find the exact failing build... But ok.

File: recipes/corrade/all/conanfile.py (hunk 2 above)
Comment (helmesjo) [reply recipes/corrade/all/conanfile.py:65]: I'd vote for `"{}".format(bool(value)).upper()`. I'm certain CMake understands `True` & `False`, but all documentation mentions uppercase only for those.

File: recipes/corrade/all/conanfile.py (hunk 1 above)
Comment (helmesjo) [reply recipes/corrade/all/conanfile.py:32]: Nah, [it's still a valid option](https://github.com/mosra/corrade/blob/61d1b58cbcb159837bca506b5336a810da67f0a7/CMakeLists.txt#L62-L65)...

File: recipes/corrade/all/conanfile.py (hunk 5 above)
Comment (helmesjo) [reply recipes/corrade/all/conanfile.py:77]: Also I could've reused `add_cmake_option(...)` here as well.

File: recipes/corrade/all/conanfile.py (hunk 2 above)
Comment (uilianries) [reply recipes/corrade/all/conanfile.py:65]: Cmake helper already converts it, don't reinvent the wheel.

File: recipes/corrade/all/conanfile.py (hunk 4 above)
Comment (uilianries) [reply recipes/corrade/all/conanfile.py:72]: keep it simple and readable like other recipes. you don't need an extra method for a simple logic.

File: recipes/corrade/all/conanfile.py (hunk 5 above)
Comment (uilianries) [reply recipes/corrade/all/conanfile.py:77]: again, don't need an extra method for a very simple logic.

File: recipes/corrade/all/conanfile.py (hunk 6 above)
Comment (uilianries) [reply recipes/corrade/all/conanfile.py:100]: I can't find a real evidence of their usage as cmake definitions. For instance, BUILD_STATIC will be "renamed" to CORRADE_BUILD_STATIC, and only the unit tests are using it. The defines in configure.h are okay, they will be processed by the consumer. Do you see a real usage for them?

https://github.com/mosra/corrade/search?q=CORRADE_UTILITY_USE_ANSI_COLORS
https://github.com/mosra/corrade/search?q=CORRADE_BUILD_STATIC

**Comment by uilianries** [comment]
Please, don't forget to sign the CLA https://github.com/conan-io/conan-center-index/pull/2109#issuecomment-653069329

**Comment by uilianries** [comment]
This package deserves components feature!

File: recipes/corrade/all/conanfile.py [hunk 8]
```diff
@@ -136,8 +147,4 @@ def package_info(self):
         if self.settings.os == "Linux":
             self.cpp_info.system_libs = ["m", "dl"]
 
-        self.cpp_info.builddirs = [os.path.join(self.package_folder, "lib", "cmake", "Corrade")]
-
-        bindir = os.path.join(self.package_folder, "bin")
-        self.output.info("Appending PATH environment variable: {}".format(bindir))
```
[the remark this answers is not in the document]
Comment (helmesjo) [reply recipes/corrade/all/conanfile.py:142]: So which is it: Keep it, or remove it..?

[the remark this answers is not in the document]
Comment (uilianries) [reply recipes/corrade/all/conanfile.py:142]: keep.

**Comment by Croydon** [comment]
The hooks message is incomplete:

> [HOOK - conan-center.py] post_package(): ERROR: [CMAKE-MODULES-CONFIG-FILES (KB-H016)] Found files:

it does not show the actual files

**Comment by Croydon** [comment]
@helmesjo You might want to close the PR for ~10 seconds and re-open it then to re-trigger CI

**Comment by tomlankhorst** [comment]
[Bad bot...](https://blog.benwinding.com/github-stale-bots/)

**Comment by danimtb** [comment]
hey @helmesjo, can you check the CLA issue? That will help us to go forward with this PR as we can make it work if you don't have the time. Thank you

**Comment by helmesjo** [comment]
> hey @helmesjo, can you check the CLA issue? That will help us to go forward with this PR as we can make it work if you don't have the time. Thank you

Just gave it one eye but can't load more comments thus can't see the state of things (can't even see my own comment)...

**Comment by danimtb** [comment]
no, everything is fine for you in #4, just check the CLA check here https://github.com/conan-io/conan-center-index/pull/2109#issuecomment-653069329

**Comment by helmesjo** [comment]
> no, everything is fine for you in #4, just check the CLA check here [#2109 (comment)](https://github.com/conan-io/conan-center-index/pull/2109#issuecomment-653069329)

Ah, forgot about that one. Alright, signed (with the assumption that it is clear that nothing with this PR violates any license provided by corrade or it's dependencies ;) ).

**Comment by tomlankhorst** [comment]
Bad bot!

**Pull request closed without merging**